File: tests/test_lesson_analysis.py
import hashlib
import json
import re
import unittest
from pathlib import Path
from xml.etree import ElementTree


ROOT = Path(__file__).resolve().parents[1]
SITE = ROOT / "site"
FIXTURE_PATH = SITE / "data" / "lesson-analysis-svg-mvp.json"
GOLDEN_PATH = SITE / "data" / "analyzer-node-k001-lesson-preview.json"
ASSET_ROOT = (
    SITE
    / "assets"
    / "positions"
    / "lesson-analysis-svg-mvp"
    / "opening-fixture"
)
CUBE_LESSON = SITE / "learn" / "cube" / "what-the-cube-is-asking.qmd"
CHECKER_LESSON = (
    SITE
    / "learn"
    / "cube"
    / "why-is-25-percent-the-basic-take-point.qmd"
)


class LessonAnalysisFixtureTests(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.data = json.loads(FIXTURE_PATH.read_text(encoding="utf-8"))
        cls.golden = json.loads(GOLDEN_PATH.read_text(encoding="utf-8"))
        cls.cube_source = CUBE_LESSON.read_text(encoding="utf-8")
        cls.checker_source = CHECKER_LESSON.read_text(encoding="utf-8")

    def test_fixture_contract_and_explicit_status(self):
        self.assertEqual(
            self.data["schema_version"],
            "bs-lesson-analysis-fixture-v1",
        )
        self.assertEqual(self.data["fixture_status"]["kind"], "fixture-only")
        message = self.data["fixture_status"]["message"].casefold()
        self.assertIn("test fixtures", message)
        self.assertIn("not verified engine output", message)

    def test_cube_fixture_supports_all_requested_answer_shapes(self):
        cube_cases = self.data["cube_cases"]
        self.assertEqual(
            cube_cases["cube-roll"]["correct_first_action"],
            "roll",
        )
        self.assertEqual(
            cube_cases["cube-double-take"]["actions"]["double"]["responder"][
                "correct_response"
            ],
            "take",
        )
        self.assertEqual(
            cube_cases["cube-double-pass"]["actions"]["double"]["responder"][
                "correct_response"
            ],
            "pass",
        )

    def test_every_referenced_svg_exists_and_parses(self):
        names = set()
        for cube in self.data["cube_cases"].values():
            names.add(cube["initial"]["image"])
            responder = cube["actions"]["double"].get("responder")
            if responder:
                names.add(responder["image"])
        for checker in self.data["checker_cases"].values():
            names.add(checker["initial"]["image"])
            names.update(candidate["image"] for candidate in checker["candidates"])

        self.assertEqual(
            names,
            {
                "starting.svg",
                "responder-flipped.svg",
                "candidate-1.svg",
                "candidate-2.svg",
                "candidate-3.svg",
            },
        )
        for name in names:
            path = ASSET_ROOT / name
            self.assertTrue(path.is_file(), name)
            root = ElementTree.parse(path).getroot()
            self.assertTrue(root.tag.endswith("svg"), name)

    def test_shared_start_is_one_asset_used_by_both_component_types(self):
        cube_start = self.data["cube_cases"]["cube-double-take"]["initial"][
            "image"
        ]
        checker_start = self.data["checker_cases"][
            "checker-three-candidates"
        ]["initial"]["image"]
        self.assertEqual(cube_start, checker_start)
        starts = list(ASSET_ROOT.rglob("starting.svg"))
        self.assertEqual(starts, [ASSET_ROOT / "starting.svg"])

    def test_real_lessons_reference_exact_golden_analysis_keys(self):
        self.assertIn(
            'data-bs-analysis-src="/data/analyzer-node-k001-lesson-preview.json"',
            self.cube_source,
        )
        self.assertIn(
            'data-bs-analysis-src="/data/analyzer-node-k001-lesson-preview.json"',
            self.checker_source,
        )
        self.assertEqual(
            self.cube_source.count("data-bs-cube-decision"),
            1,
        )
        self.assertEqual(
            self.checker_source.count("data-bs-checker-decision"),
            1,
        )
        self.assertIn(
            "sha256-52e8ef0da2e4090a81f0ab726370811812c20f76f31730c5e6d132e63b774f3d",
            self.checker_source,
        )
        self.assertIn(
            "sha256-1217f65d4a2c203e2370edb860ffaba81090a42f69d2a5fb56f5cceb64389e01",
            self.cube_source,
        )
        self.assertNotIn("<svg", self.cube_source.casefold())
        self.assertNotIn("<svg", self.checker_source.casefold())

    def test_golden_checker_preserves_source_candidates_and_prepared_assets(self):
        checker = self.golden["analyses"][
            "sha256-52e8ef0da2e4090a81f0ab726370811812c20f76f31730c5e6d132e63b774f3d"
        ]
        self.assertEqual(checker["metadata"]["recommendation"], "8/4 6/4")
        self.assertEqual(len(checker["candidates"]), 8)
        self.assertEqual(
            [candidate["source_order"] for candidate in checker["candidates"]],
            list(range(1, 9)),
        )
        self.assertEqual(
            [candidate["id"] for candidate in checker["candidates"]],
            [f"gnu-move-{rank}" for rank in range(1, 9)],
        )
        for rank, candidate in enumerate(checker["candidates"], start=1):
            expected = f"/assets/positions/node-k001/checker/candidate-{rank}.svg"
            self.assertEqual(candidate["move_board"]["image"], expected)
            self.assertTrue((SITE / expected.removeprefix("/")).is_file())

    def test_golden_cube_preserves_source_actions_equities_and_probabilities(self):
        cube = self.golden["analyses"][
            "sha256-1217f65d4a2c203e2370edb860ffaba81090a42f69d2a5fb56f5cceb64389e01"
        ]
        self.assertEqual(cube["metadata"]["recommendation"], "Double, take")
        self.assertEqual(
            [(action["id"], action["value"]["value"]) for action in cube["actions"]],
            [
                ("double-take", 0.998032),
                ("double-pass", 1.0),
                ("no-double", 0.637873),
            ],
        )
        self.assertEqual(
            cube["probabilities"],
            {
                "win": 0.749996,
                "win_gammon_or_better": 0.0,
                "win_backgammon": 0.0,
                "lose": 0.250004,
                "lose_gammon_or_worse": 0.0,
                "lose_backgammon": 0.0,
            },
        )
        self.assertTrue(all(action["probabilities"] is None for action in cube["actions"]))

    def test_qmd_hosts_do_not_hard_code_component_ids(self):
        for source in (self.cube_source, self.checker_source):
            host_blocks = re.findall(
                r"<div\s+.*?data-bs-(?:cube|checker)-decision.*?</div>",
                source,
                flags=re.DOTALL,
            )
            self.assertTrue(host_blocks)
            for block in host_blocks:
                self.assertNotRegex(block, r'(?:^|\s)id="')

    def test_script_is_loaded_before_continuous_lesson_loader(self):
        scripts = (SITE / "includes" / "bs-scripts.html").read_text(
            encoding="utf-8"
        )
        analysis_index = scripts.index("bs-lesson-analysis.js")
        shared_index = scripts.index("bs-analysis-results.js")
        scroll_index = scripts.index("bs-learn-scroll.js")
        self.assertLess(shared_index, analysis_index)
        self.assertLess(analysis_index, scroll_index)
        implementation = (
            SITE / "assets" / "bs-lesson-analysis.js"
        ).read_text(encoding="utf-8")
        self.assertIn("window.BSLearn.mountLesson", implementation)
        self.assertIn("mount(rootElement)", implementation)
        self.assertIn("dataset.bsAnalysisMounted", implementation)

    def test_svg_reuse_cannot_duplicate_inline_ids(self):
        implementation = (
            SITE / "assets" / "bs-lesson-analysis.js"
        ).read_text(encoding="utf-8")
        self.assertIn("sharedAnalysis().renderBoard", implementation)
        self.assertNotIn('document.createElement("img")', implementation)
        self.assertNotIn("fetchSvg", implementation)
        start_hash = hashlib.sha256(
            (ASSET_ROOT / "starting.svg").read_bytes()
        ).hexdigest()
        self.assertEqual(len(start_hash), 64)

    def test_missing_optional_values_are_retained(self):
        candidate = self.data["checker_cases"]["checker-three-candidates"][
            "candidates"
        ][2]
        self.assertIsNone(candidate["winning_probabilities"]["win_gammon"])
        self.assertIsNone(candidate["winning_probabilities"]["lose_gammon"])

    def test_resource_contract_copies_dynamic_assets(self):
        config = (SITE / "_quarto.yml").read_text(encoding="utf-8")
        self.assertIn('"assets/positions/**"', config)
        self.assertIn("data/lesson-analysis-svg-mvp.json", config)
        self.assertIn("data/checker-sage-gnu-disagreement-001.json", config)
        self.assertIn("data/analyzer-node-k001-lesson-preview.json", config)
        self.assertIn("assets/bs-lesson-analysis.css", config)
        provenance = ASSET_ROOT / "PROVENANCE.txt"
        self.assertTrue(provenance.is_file())
        self.assertFalse((ASSET_ROOT / "PROVENANCE.md").exists())

    def test_lesson_is_a_thin_adapter_to_shared_presentation(self):
        implementation = (
            SITE / "assets" / "bs-lesson-analysis.js"
        ).read_text(encoding="utf-8")
        shared = (SITE / "assets" / "bs-analysis-results.js").read_text(
            encoding="utf-8"
        )
        lesson_css = (SITE / "assets" / "bs-lesson-analysis.css").read_text(
            encoding="utf-8"
        )
        self.assertIn("checkerViewModel", implementation)
        self.assertIn("cubeViewModel", implementation)
        self.assertIn("sharedAnalysis().renderPresentation", implementation)
        self.assertIn(".fixtureLoader(analysisUrl)(analysisId)", implementation)
        self.assertIn("mountAcceptedChecker", implementation)
        self.assertIn("mountAcceptedCube", implementation)
        self.assertNotIn("candidateMetricRows", implementation)
        self.assertNotIn("analysisRows", implementation)
        self.assertNotIn(".bs-analysis-position-image", lesson_css)
        self.assertNotIn(".bs-analysis-metrics", lesson_css)
        self.assertIn("function outcomePanel", shared)
        self.assertIn("function renderChecker", shared)
        self.assertIn("function renderCube", shared)

    def test_real_lesson_flow_remains_owned_by_the_adapter(self):
        implementation = (
            SITE / "assets" / "bs-lesson-analysis.js"
        ).read_text(encoding="utf-8")
        self.assertIn('choiceButton(candidate.move, candidate.id)', implementation)
        self.assertIn('choiceButton("Double", "double")', implementation)
        self.assertIn('choiceButton("Take", "take")', implementation)
        self.assertIn("revealAcceptedAnalysis", implementation)
        self.assertIn("data-bs-lesson-prompt", self.checker_source)
        self.assertIn("data-bs-lesson-prompt", self.cube_source)

    def test_direct_browser_path_does_not_transform_or_calculate_analysis(self):
        implementation = (
            SITE / "assets" / "bs-lesson-analysis.js"
        ).read_text(encoding="utf-8")
        direct_path = implementation.split(
            "function acceptedAnalysisChoice", 1
        )[1].split("function mountCube", 1)[0]
        self.assertIn("payload.analysis", implementation)
        for forbidden in (
            "Parquet",
            "DuckDB",
            "GNUID",
            "Position ID",
            "decode",
            "applyMove",
            "parseMove",
            "difference_from_best",
            "equity_loss",
        ):
            self.assertNotIn(forbidden, direct_path)


if __name__ == "__main__":
    unittest.main()

File: tests/test_lesson_analysis.js
const assert = require("node:assert/strict");
const fs = require("node:fs");
const path = require("node:path");

const analysis = require("../site/assets/bs-lesson-analysis.js");
const shared = require("../site/assets/bs-analysis-results.js");

const root = path.resolve(__dirname, "..");
const fixtures = JSON.parse(
  fs.readFileSync(
    path.join(root, "site/data/lesson-analysis-svg-mvp.json"),
    "utf8"
  )
);
const realFixtures = JSON.parse(
  fs.readFileSync(
    path.join(root, "site/data/checker-sage-gnu-disagreement-001.json"),
    "utf8"
  )
);
const goldenAnalyses = JSON.parse(
  fs.readFileSync(
    path.join(root, "site/data/analyzer-node-k001-lesson-preview.json"),
    "utf8"
  )
);

assert.equal(
  analysis.validateFixtureDocument(fixtures),
  fixtures,
  "the checked-in fixture document is accepted"
);
assert.equal(analysis.validateFixtureDocument(realFixtures), realFixtures);

const goldenChecker = shared.analysisFromDocument(
  goldenAnalyses,
  "sha256-52e8ef0da2e4090a81f0ab726370811812c20f76f31730c5e6d132e63b774f3d"
).analysis;
const goldenCube = shared.analysisFromDocument(
  goldenAnalyses,
  "sha256-1217f65d4a2c203e2370edb860ffaba81090a42f69d2a5fb56f5cceb64389e01"
).analysis;
assert.equal(goldenChecker.metadata.recommendation, "8/4 6/4");
assert.equal(goldenChecker.candidates.length, 8);
assert.deepEqual(
  goldenChecker.candidates.map((candidate) => candidate.source_order),
  [1, 2, 3, 4, 5, 6, 7, 8]
);
assert.equal(
  analysis.acceptedAnalysisChoice(goldenChecker, "gnu-move-8").move,
  "24/20 6/4"
);
assert.match(
  analysis.acceptedAnalysisChoice(goldenChecker, "gnu-move-8").move_board.image,
  /node-k001\/checker\/candidate-8\.svg$/
);
assert.equal(goldenCube.metadata.recommendation, "Double, take");
assert.match(
  goldenCube.original_board.image,
  /node-k001\/cube\/starting\.svg$/
);
assert.match(
  goldenCube.responder_board.image,
  /node-k001\/cube\/responder\.svg$/
);
assert.deepEqual(
  goldenCube.actions.map((action) => [action.id, action.value.value]),
  [
    ["double-take", 0.998032],
    ["double-pass", 1.0],
    ["no-double", 0.637873]
  ]
);
assert.equal(
  analysis.acceptedAnalysisChoice(goldenCube, "double-take").probabilities,
  null
);
assert.equal(goldenCube.context.decision, "Cube decision");
assert.equal(goldenCube.context.dice, null);
assert.throws(
  () => analysis.acceptedAnalysisChoice(goldenCube, "roll"),
  /does not define/
);

const doubleTake = fixtures.cube_cases["cube-double-take"];
const rollReview = analysis.cubeDecisionState(doubleTake, "roll");
assert.equal(rollReview.actionAccepted, false);
assert.equal(rollReview.responder, null);
assert.equal(rollReview.actionData.analysis.recommendation, "Double");

const doubleChoice = analysis.cubeDecisionState(doubleTake, "double");
assert.equal(doubleChoice.actionAccepted, true);
assert.equal(doubleChoice.responder.correct_response, "take");
assert.equal(doubleChoice.actionData.analysis.recommendation, "Double");

const passReview = analysis.cubeDecisionState(doubleTake, "double", "pass");
assert.equal(passReview.responseAccepted, false);
assert.equal(passReview.responseData.analysis.recommendation, "Take");

const takeAnswer = analysis.cubeDecisionState(doubleTake, "double", "take");
assert.equal(takeAnswer.responseAccepted, true);
assert.equal(takeAnswer.responseData.analysis.recommendation, "Take");

const rollFixture = fixtures.cube_cases["cube-roll"];
assert.equal(
  analysis.cubeDecisionState(rollFixture, "roll").actionAccepted,
  true,
  "the component supports Roll as the accepted first action"
);
assert.equal(
  analysis.cubeDecisionState(rollFixture, "double").responder,
  null,
  "a rejected Double does not invent a responder decision"
);

const doublePass = fixtures.cube_cases["cube-double-pass"];
assert.equal(
  analysis.cubeDecisionState(doublePass, "double", "pass").responseAccepted,
  true,
  "the component supports Pass as the accepted cube response"
);
assert.equal(
  analysis.cubeDecisionState(doublePass, "double", "take").responseAccepted,
  false
);

assert.throws(
  () => analysis.cubeDecisionState(doubleTake, "beaver"),
  /Double or Roll/
);

const realChecker =
  realFixtures.checker_cases["checker-sage-gnu-disagreement-001"];
const realCandidate = analysis.checkerCandidateState(
  realChecker,
  "candidate-1"
);
assert.equal(realCandidate.label, "8/4");
assert.equal(realCandidate.rank, 1);
assert.equal(realCandidate.equity, -1.615);
assert.equal(realCandidate.equity_loss, 0);
assert.equal(realCandidate.winning_probabilities.lose_gammon, 0.677);
assert.equal(realCandidate.explanation, null);
assert.equal(
  analysis.checkerCandidateIdentityMatches(realChecker, realCandidate),
  true
);
assert.equal(
  analysis.checkerCandidateIdentityMatches(realChecker, {
    ...realCandidate,
    analysis_id: "wrong"
  }),
  false
);
assert.throws(
  () => analysis.cubeDecisionState(doubleTake, "double", "beaver"),
  /Pass or Take/
);

const checker = fixtures.checker_cases["checker-three-candidates"];
const candidate1 = analysis.checkerCandidateState(checker, "candidate-1");
const candidate2 = analysis.checkerCandidateState(checker, "candidate-2");
const candidate3 = analysis.checkerCandidateState(checker, "candidate-3");
assert.equal(candidate1.image, "candidate-1.svg");
assert.equal(candidate2.equity_loss, 0);
assert.equal(candidate3.winning_probabilities.win_gammon, null);
assert.throws(
  () => analysis.checkerCandidateState(checker, "candidate-4"),
  /does not define/
);

const checkerModel = analysis.checkerViewModel(checker, fixtures);
assert.equal(checkerModel.analysis_kind, "checker");
assert.equal(checkerModel.candidates.length, 3);
assert.equal(checkerModel.candidates[1].value.value, 0.093);
assert.equal(checkerModel.candidates[1].difference_from_best, -0);
assert.equal(checkerModel.candidates[2].probabilities.win_gammon_or_better, null);
assert.match(checkerModel.candidates[0].move_board.image, /candidate-1\.svg$/);
assert.equal(shared.validateAnalysisModel(checkerModel), checkerModel);

const retainedCheckerModel = analysis.checkerViewModel(realChecker, realFixtures);
assert.equal(retainedCheckerModel.candidates[0].move, "8/4");
assert.equal(retainedCheckerModel.candidates[1].difference_from_best, -0.002);
assert.equal(
  retainedCheckerModel.candidates[0].probabilities.lose_gammon_or_worse,
  0.677
);
assert.match(
  retainedCheckerModel.candidates[0].move_board.alt,
  /starting position with checker movement overlay/
);
assert.equal(shared.validateAnalysisModel(retainedCheckerModel), retainedCheckerModel);

const cubeModel = analysis.cubeViewModel(
  doubleTake.actions.double.analysis,
  doubleTake.initial,
  doubleTake,
  fixtures,
  "first-double"
);
assert.equal(cubeModel.analysis_kind, "cube");
assert.deepEqual(
  cubeModel.actions.map((action) => action.id),
  ["roll", "double_take"]
);
assert.equal(cubeModel.probabilities.lose, 0.37);
assert.equal(cubeModel.probabilities.win_gammon_or_better, 0.14);
assert.equal(analysis.matchingActionId(cubeModel, "double"), "double_take");
assert.equal(shared.validateAnalysisModel(cubeModel), cubeModel);

assert.deepEqual(analysis.lessonProbabilities({ win: 0.58 }), {
  win: 0.58,
  win_gammon_or_better: undefined,
  win_backgammon: undefined,
  lose: 0.42000000000000004,
  lose_gammon_or_worse: undefined,
  lose_backgammon: undefined
});

assert.equal(
  analysis.assetUrl(fixtures.asset_root, fixtures.cube_cases["cube-roll"].initial.image),
  "/assets/positions/lesson-analysis-svg-mvp/opening-fixture/starting.svg"
);
assert.throws(
  () => analysis.assetUrl(fixtures.asset_root, "../outside.svg"),
  /unsafe/
);

analysis.resetInstanceCounter();
const instanceIds = [
  analysis.nextInstanceId("cube", "cube-double-take"),
  analysis.nextInstanceId("cube", "cube-double-take"),
  analysis.nextInstanceId("checker", "checker-three-candidates")
];
assert.equal(new Set(instanceIds).size, instanceIds.length);
assert.match(instanceIds[0], /^bs-analysis-cube-cube-double-take-1$/);
assert.match(instanceIds[2], /^bs-analysis-checker-checker-three-candidates-3$/);

console.log("lesson analysis fixture logic passed");
